### ISSUE TITLE
vfs: file handle leak with --vfs-cache-mode full and --buffer-size 0

### ISSUE THREAD

**Issue by ivandeex** [rclone/rclone#4892]
Originally posted by  @daniele-spl in https://github.com/rclone/rclone/issues/2614#issuecomment-747559631, https://github.com/rclone/rclone/issues/2614#issuecomment-748204966

#### What is the problem you are having with rclone?

lsof output right now, application already closed some files
lsof -c rclone | wc -l
1027

Error 1: vfs cache: failed to transfer file from cache to remote: failed to open source object .... too many open files

Error 2: Non-out-of-space error encountered during open

As soon as this happens, i can see RSS of rclone is increasing to some gig and CPU is up to 400%. Normal mode is 20/30% CPU.

The open file issue is just a result of the the main issue. I mean, a static limit to somewhere near 1030 anyhow is not that good, but in most cases you shouldnt come up there.

My issue was, that vfs caching did not work properly.

I get the same error like #4639 

Its really strange. After a reboot of the system I get:
```
2020/12/18 15:40:14 ERROR : Failed to create vfs cache - disabling: failed to make cache directory: make cache directory failed: mkdir /mnt/resource/rclonecache: permission denied
```

After that its doing a direct transfer to azure. Cause I open a lot of files and also a lot of small files. So the open file issue is related to the http connections to the azure storage.

Then you wrote me to enable debug log, so I restarted the service with the debug flag, It restarted without any errors. The above Error does not come up, when i restart the service or just start it without the deamon flag at command line.
BUT: It still cannot use the cache and is creating interesting logs but does nothing at all.

#### What is your rclone version, OS, backend

Version: rclone-1.53.2
OS: SLES 12 SP5
Backend: Azure Blob

#### The command you were trying to run (e.g. `rclone copy /tmp remote:tmp`)

`rclone mount`

I did a `cp` from the mount to a local directory

I found the issue at least for me. I had the flag --buffer-size 0
Thats not working and caused all the error messages during transfer with cache configured.

These are my current testing parameters:
--cache-dir /tmp/cache --track-renames --dir-cache-time 24h --attr-timeout 4m --vfs-write-back 5s --vfs-cache-mode=full --checksum --vfs-cache-max-age 1m --vfs-cache-max-size 200G --log-level=DEBUG --allow-other --allow-non-empty --debug-fuse --buffer-size 1

The same with buffer-size 0 is throwing the errors.

#### A log from the command with the `-vv` flag (e.g. output from `rclone -vv copy /tmp remote:tmp`)

<details>
<summary>Click to see log</summary>

```
2020/12/18 16:37:30 DEBUG : test/journal.gz3: ChunkedReader.Read at 819200 length 32768 chunkOffset 786432 chunkSize 134217728
2020/12/18 16:37:30 DEBUG : test/journal.gz3: ChunkedReader.Read at -1 length 32768 chunkOffset 1572864 chunkSize 134217728
2020/12/18 16:37:30 DEBUG : test/journal.gz3: ChunkedReader.openRange at 1572864 length 134217728
2020/12/18 16:37:30 DEBUG : test/journal.gz3: ChunkedReader.RangeSeek from -1 to 1310720 length -1
2020/12/18 16:37:30 DEBUG : test/journal.gz3: ChunkedReader.Read at -1 length 32768 chunkOffset 1310720 chunkSize 134217728
2020/12/18 16:37:30 DEBUG : test/journal.gz3: ChunkedReader.openRange at 1310720 length 134217728
2020/12/18 16:37:30 DEBUG : test/journal.gz3: ChunkedReader.RangeSeek from -1 to 1441792 length -1
2020/12/18 16:37:30 DEBUG : test/journal.gz3: ChunkedReader.Read at -1 length 32768 chunkOffset 1441792 chunkSize 134217728
2020/12/18 16:37:30 DEBUG : test/journal.gz3: ChunkedReader.openRange at 1441792 length 134217728
2020/12/18 16:37:30 DEBUG : test/journal.gz3: ChunkedReader.RangeSeek from -1 to 1703936 length -1
2020/12/18 16:37:30 DEBUG : test/journal.gz3: ChunkedReader.Read at -1 length 32768 chunkOffset 1703936 chunkSize 134217728
2020/12/18 16:37:30 DEBUG : test/journal.gz3: ChunkedReader.openRange at 1703936 length 134217728
2020/12/18 16:37:30 DEBUG : test/journal.gz3: ChunkedReader.RangeSeek from -1 to 1572864 length -1
2020/12/18 16:37:30 DEBUG : test/journal.gz3: ChunkedReader.Read at 851968 length 32768 chunkOffset 786432 chunkSize 134217728
2020/12/18 16:37:30 DEBUG : test/journal.gz3: ChunkedReader.Read at -1 length 32768 chunkOffset 1572864 chunkSize 134217728
2020/12/18 16:37:30 DEBUG : test/journal.gz3: ChunkedReader.openRange at 1572864 length 134217728
2020/12/18 16:37:30 DEBUG : test/journal.gz3: ChunkedReader.RangeSeek from -1 to 1310720 length -1
2020/12/18 16:37:30 DEBUG : test/journal.gz3: ChunkedReader.Read at -1 length 32768 chunkOffset 1310720 chunkSize 134217728
2020/12/18 16:37:30 DEBUG : test/journal.gz3: ChunkedReader.openRange at 1310720 length 134217728
2020/12/18 16:37:30 DEBUG : test/journal.gz3: ChunkedReader.RangeSeek from -1 to 1441792 length -1
2020/12/18 16:37:30 DEBUG : test/journal.gz3: ChunkedReader.Read at -1 length 32768 chunkOffset 1441792 chunkSize 134217728
2020/12/18 16:37:30 DEBUG : test/journal.gz3: ChunkedReader.openRange at 1441792 length 134217728
2020/12/18 16:37:30 DEBUG : test/journal.gz3: ChunkedReader.RangeSeek from -1 to 1703936 length -1
2020/12/18 16:37:30 DEBUG : test/journal.gz3: ChunkedReader.Read at -1 length 32768 chunkOffset 1703936 chunkSize 134217728
2020/12/18 16:37:30 DEBUG : test/journal.gz3: ChunkedReader.openRange at 1703936 length 134217728
2020/12/18 16:37:30 DEBUG : test/journal.gz3: ChunkedReader.RangeSeek from -1 to 1572864 length -1
2020/12/18 16:37:30 DEBUG : test/journal.gz3: ChunkedReader.Read at 884736 length 32768 chunkOffset 786432 chunkSize 134217728
2020/12/18 16:37:30 DEBUG : test/journal.gz3: ChunkedReader.Read at -1 length 32768 chunkOffset 1572864 chunkSize 134217728
2020/12/18 16:37:30 DEBUG : test/journal.gz3: ChunkedReader.openRange at 1572864 length 134217728
2020/12/18 16:37:30 DEBUG : test/journal.gz3: ChunkedReader.RangeSeek from -1 to 1310720 length -1
2020/12/18 16:37:30 DEBUG : test/journal.gz3: ChunkedReader.Read at -1 length 32768 chunkOffset 1310720 chunkSize 134217728
2020/12/18 16:37:30 DEBUG : test/journal.gz3: ChunkedReader.openRange at 1310720 length 134217728
2020/12/18 16:37:30 DEBUG : test/journal.gz3: ChunkedReader.RangeSeek from -1 to 1441792 length -1
2020/12/18 16:37:30 DEBUG : test/journal.gz3: ChunkedReader.Read at -1 length 32768 chunkOffset 1441792 chunkSize 134217728
2020/12/18 16:37:30 DEBUG : test/journal.gz3: ChunkedReader.openRange at 1441792 length 134217728
2020/12/18 16:37:30 DEBUG : test/journal.gz3: ChunkedReader.RangeSeek from -1 to 1703936 length -1
2020/12/18 16:37:30 DEBUG : test/journal.gz3: ChunkedReader.Read at -1 length 32768 chunkOffset 1703936 chunkSize 134217728
2020/12/18 16:37:30 DEBUG : test/journal.gz3: ChunkedReader.openRange at 1703936 length 134217728
2020/12/18 16:37:30 DEBUG : test/journal.gz3: ChunkedReader.RangeSeek from -1 to 1572864 length -1
2020/12/18 16:37:30 DEBUG : test/journal.gz3: ChunkedReader.Read at 917504 length 32768 chunkOffset 786432 chunkSize 134217728
2020/12/18 16:37:30 DEBUG : test/journal.gz3: ChunkedReader.Read at -1 length 32768 chunkOffset 1572864 chunkSize 134217728
2020/12/18 16:37:30 DEBUG : test/journal.gz3: ChunkedReader.openRange at 1572864 length 134217728
```

</details>

**Comment by ncw** [comment]
The problem seems to be the specific combination of `--vfs-cache-mode full` and `--buffer-size 0` when using `rclone mount`

I managed to reproduce this

```
rclone mount -vv --buffer-size 0 --vfs-cache-mode full TestDrive: /mnt/tmp/
```

And doing this on an uncached file

```
md5sum /mnt/tmp/10M
```

This seems to open more and more downloaders which explains the symptoms.

@daniele-spl can you give this a go and make sure it fixes the problem for you too please?

[v1.54.0-beta.5012.f8e9bd123.fix-4892-vfs-window](https://beta.rclone.org/branch/fix-4892-vfs-window/v1.54.0-beta.5012.f8e9bd123.fix-4892-vfs-window/) on branch [fix-4892-vfs-window](https://github.com/rclone/rclone/tree/fix-4892-vfs-window) (uploaded in 15-30 mins)

**Comment by daniele-spl** [comment]
Sorry for the late response.
I tried that out and have an error starting rclone.

2021/01/12 10:40:36 DEBUG : rclone: Version "v1.54.0-beta.5012.f8e9bd123.fix-4892-vfs-window" starting with parameters ["/usr/bin/rclone" "mount" "xxx" "xxx" "--daemon" "--cache-dir" "/mnt/resource/rclonecache" "--track-renames" "--dir-cache-time" "24h" "--attr-timeout" "4m" "--vfs-write-back" "5s" "--vfs-cache-mode" "full" "--transfers=16" "--azureblob-upload-cutoff" "32m" "--azureblob-chunk-size" "32m" "--checksum" "--vfs-read-chunk-size" "32m" "--vfs-read-chunk-size-limit" "500m" "--multi-thread-cutoff=32m" "--multi-thread-streams=16" "--vfs-cache-max-age" "168h" "--vfs-cache-max-size" "300g" "--log-level=DEBUG" "--allow-other" "--allow-non-empty" "--use-mmap" "--buffer-size" "0"]
2021/01/12 10:40:36 DEBUG : Creating backend with remote "xxx"
2021/01/12 10:40:36 DEBUG : Using config file from "/opt/xxx/.rclone.conf"
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0xadb5fa]

goroutine 1 [running]:
github.com/rclone/rclone/backend/azureblob.NewFs(0x1e6f400, 0xc000128000, 0x7fff99084cea, 0xb, 0x7fff99084cf6, 0x14, 0x1e4bb00, 0xc0000fb200, 0x0, 0x0, ...)
        github.com/rclone/rclone/backend/azureblob/azureblob.go:652 +0x19da
github.com/rclone/rclone/fs.NewFs(0x1e6f400, 0xc000128000, 0x7fff99084cea, 0x20, 0xb, 0xffffffffffffffff, 0x39, 0x8)
        github.com/rclone/rclone/fs/fs.go:1366 +0x154
github.com/rclone/rclone/fs/cache.GetFn.func1(0x7fff99084cea, 0x20, 0x7fff99084cea, 0x20, 0x2954ca0, 0x0, 0x0)
        github.com/rclone/rclone/fs/cache/cache.go:48 +0x5b
github.com/rclone/rclone/lib/cache.(*Cache).Get(0xc000122f90, 0x7fff99084cea, 0x20, 0xc00049fbf0, 0x0, 0x0, 0x0, 0x0)
        github.com/rclone/rclone/lib/cache/cache.go:61 +0x150
github.com/rclone/rclone/fs/cache.GetFn(0x1e6f400, 0xc000128000, 0x7fff99084cea, 0x20, 0x1c28b28, 0x2922440, 0x0, 0x32, 0x2954ca0)
        github.com/rclone/rclone/fs/cache/cache.go:47 +0xfe
github.com/rclone/rclone/fs/cache.Get(...)
        github.com/rclone/rclone/fs/cache/cache.go:104
github.com/rclone/rclone/cmd.newFsDir(0x7fff99084cea, 0x20, 0xc0001fedc0, 0xc0005e9b80)
        github.com/rclone/rclone/cmd/cmd.go:142 +0x71
github.com/rclone/rclone/cmd.NewFsDir(...)
        github.com/rclone/rclone/cmd/cmd.go:155
github.com/rclone/rclone/cmd/mountlib.NewMountCommand.func1(0xc0001fedc0, 0xc0005e9b80, 0x2, 0x25)
        github.com/rclone/rclone/cmd/mountlib/mount.go:437 +0x14b
github.com/spf13/cobra.(*Command).execute(0xc0001fedc0, 0xc0005e9900, 0x25, 0x28, 0xc0001fedc0, 0xc0005e9900)
        github.com/spf13/cobra@v1.1.1/command.go:854 +0x2c2
github.com/spf13/cobra.(*Command).ExecuteC(0x29050c0, 0x0, 0x4062c5, 0xc000114058)
        github.com/spf13/cobra@v1.1.1/command.go:958 +0x375
github.com/spf13/cobra.(*Command).Execute(...)
        github.com/spf13/cobra@v1.1.1/command.go:895
github.com/rclone/rclone/cmd.Main()
        github.com/rclone/rclone/cmd/cmd.go:538 +0xaa
main.main()
        github.com/rclone/rclone/rclone.go:14 +0x25

**Comment by daniele-spl** [comment]
Is the buffer-size also handling the Buffer of the azureblob upload buffer or is there another flag for this?

**Comment by ncw** [comment]
Sorry this is an unrelated bug which is now fixed. I rebased the beta which should hopefully fix the problem

[v1.54.0-beta.5062.ba5657332.fix-4892-vfs-window](https://beta.rclone.org/branch/fix-4892-vfs-window/v1.54.0-beta.5062.ba5657332.fix-4892-vfs-window/) on branch [fix-4892-vfs-window](https://github.com/rclone/rclone/tree/fix-4892-vfs-window) (uploaded in 15-30 mins)

**Comment by daniele-spl** [comment]
It seems to work properly in this version. I can see the stuff in the cache directories also with buffer size 0 and do not get any errors.
Great work. Thanks!

**Comment by ncw** [comment]
Thanks for testing :-)

I've merged this to master now which means it will be in [the latest beta](https://beta.rclone.org) in 15-30 mins and released in v1.54
